doc/database/multi_database_instances.md

```diff
@@ -820,36 +820,193 @@ EXEC_WITH_ERROR_THROW(cmd, res);
+  	          cursor = dat[1];
+              redis.call('MIGRATE', KEYS[1], KEYS[2], '', KEYS[3], 5000, 'REPLACE', 'KEYS', unpack(dat[2]));
+          until cursor == '0';
+      """
+      r.eval(script, 3, '127.0.0.1', 6381, dbid)
+
+  ```
+  
+  we can loop on instances as well and operate each database inside lua script, the perfoemacne is the same as we did in the above script.
+  
+  ```python
+  #!/usr/bin/python
+  from __future__ import print_function
+  import sys
+  import swsssdk
+  import redis
+  
+  instlists = swsssdk.SonicDBConfig.get_instancelist()
+  for k, v in instlists.items():
+      instsocket = v['unix_socket_path']
+      r = redis.Redis(unix_socket_path=instsocket)
+  
+      script = """
+          local lines = redis.call('INFO', 'keyspace')
+          for line in lines:gmatch("([^\\n\\r]+)") do
+              local dbid = line:match("(%d+)")
+              if dbid ~= nil then
+                  local cursor = 0;
+                  redis.call('SELECT', dbid)
+                  repeat
+                      local  dat = redis.call('SCAN', cursor, 'COUNT', 7000);
+                      cursor = dat[1];
+                      redis.call('MIGRATE', KEYS[1], KEYS[2], '', dbid, 5000, 'COPY', 'REPLACE', 'KEYS', unpack(dat[2]));
+                  until cursor == '0';
+              end
+          end
+      """
+      r.eval(script, 2, '127.0.0.1', 6382)
+  ```
+  
+  Of course, we can preload the lua script to redis server to improve the performance, but from the test result, it is not helping, almost the same performance as we don't preload since the lua script is very simple and short.
+  
+  
+  
+  For example, migrating below four instances into one instance, total **data size ~100K, costs ~2.1s**. TCP and UNIX SOCKET connections almost take the same time.
+  
+  ```shell
+  admin@ASW-7005:~$ redis-cli -p 6379 info | grep Keyspace -A 10
+  # Keyspace
+  db0:keys=40012,expires=0,avg_ttl=0
+  db4:keys=99,expires=0,avg_ttl=0
+  db5:keys=3145,expires=0,avg_ttl=0
+  db6:keys=365,expires=0,avg_ttl=0
+  
+  admin@ASW-7005:~$ redis-cli -p 6380 info | grep Keyspace -A 10
+  # Keyspace
+  db1:keys=42834,expires=0,avg_ttl=0
+  
+  admin@ASW-7005:~$ redis-cli -p 6381 info | grep Keyspace -A 10
+  # Keyspace
+  
+  admin@ASW-7005:~$ redis-cli -p 6382 info | grep Keyspace -A 10
+  # Keyspace
+  db2:keys=7191,expires=0,avg_ttl=0
+  
+  admin@ASW-7005:~$ redis-cli -p 6383 info | grep Keyspace -A 10
+  # Keyspace
+  db3:keys=103,expires=0,avg_ttl=0
+  
+  admin@ASW-7005:~$ date +"%T.%N" ; sudo /etc/sonic/migratedb ;  date +"%T.%N"
+  23:15:08.350089582
+  23:15:10.486322265
+  
+  admin@ASW-7005:~$ redis-cli -p 6381 info | grep Keyspace -A 10
+  # Keyspace
+  db0:keys=40012,expires=0,avg_ttl=0
+  db1:keys=42834,expires=0,avg_ttl=0
+  db2:keys=7191,expires=0,avg_ttl=0
+  db3:keys=103,expires=0,avg_ttl=0
+  db4:keys=99,expires=0,avg_ttl=0
+  db5:keys=3145,expires=0,avg_ttl=0
+  db6:keys=365,expires=0,avg_ttl=0
+  ```
```

Comment:
In this output, find that the "DB names" are used to identify the databases after migration into a single instance, eg: db0, db1 etc ? 
If that is the case , even with current single database docker scheme -- we need to ensure that the database names are unique across different redis instances. eg: COUNTERS_DB cannot be there in redis instance0 and redis instance1.

Reply:
in single DB case, today  database names are unique across different redis instance, this is why this approach is proposed.

Reply:
Sure, thanks - it should be ok then ! I am wondering if we need to add some validation to enforce this in SonicDBConfig or so when we parse the database_config file ?